src/mlservice.h

```diff
@@ -495,36 +499,29 @@ namespace dd
      */
     int predict_job(const APIData &ad, APIData &out, const bool &chain = false)
     {
-      // TODO: collect input transformed data for chain, store it here in
-      // memory
-      // -> beware, the input connector is a copy...
+      if (!_train_mutex.try_lock_shared())
+        throw MLServiceLockException(
+            "Predict call while training with an offline learning algorithm");
 
-      if (!this->_online)
+      this->_stats.predict_start();
+
+      int err = 0;
+      try
         {
-          if (!_train_mutex.try_lock_shared())
-            throw MLServiceLockException("Predict call while training with an "
-                                         "offline learning algorithm");
-          int err = 0;
-          try
-            {
-              if (chain)
-                const_cast<APIData &>(ad).add("chain", true);
-              err = this->predict(ad, out);
-            }
-          catch (std::exception &e)
-            {
-              _train_mutex.unlock_shared();
-              throw;
-            }
-          _train_mutex.unlock_shared();
-          return err;
+          if (chain)
+            const_cast<APIData &>(ad).add("chain", true);
+          err = this->predict(ad, out);
         }
-      else // wait til a lock can be acquired
+      catch (std::exception &e)
         {
-          boost::shared_lock<boost::shared_mutex> lock(_train_mutex);
-          return this->predict(ad, out);
+          _train_mutex.unlock_shared();
+          this->_stats.predict_end(false);
+          throw;
         }
-      return 0;
+      this->_stats.predict_end(true);
+
+      _train_mutex.unlock_shared();
```

Comment:
Just curious, we need to explicitely unlock the shared mutex ?

Reply:
I just keep how you rewrote this method in your branch and it doesn't use `std::lock_guard`. I guess this is because we use `unlock_shared()` to unlock it. If we use`std::lock_guard`, it will call `unlock()`  instead and this is not want we want here.